blog/_posts/2017-02-05-offset-arrays.md

```diff
@@ -0,0 +1,489 @@
+---
+layout: post
+title:  "Now we know where we've been: custom array indices in Julia"
+author: <a href="http://holylab.wustl.edu">Tim Holy</a>
+---
+
+# Introduction: Arrays and indices
+
+Arrays are a crucial component of any programming language,
+particularly for a data-oriented language like Julia.  Arrays store
+values according to their location: in Julia, given a two-dimensional
+array `A`, the expression `A[1,3]` returns the value stored at a
+location known as `(1,3)`.  If, for example, `A` stores `Float64`
+numbers, the value returned by this expression will be a single
+`Float64` number.
+
+Julia's arrays conventionally start numbering their axes with 1,
+meaning that the first element of a one-dimensional array `a` is
+`a[1]`. The choice of 1 vs. 0 seems to generate a certain amount of
+discussion. A fairly recent addition to the Julia landscape is the
+ability to define arrays that start with an *arbitrary* index.  The
+purpose of this blog post is to describe why this might be
+interesting.  This is really a "user-oriented" blog post, hinting at
+some of the ways this new feature can make your life simpler.  For
+developers who want to write code that supports arrays with arbitrary
+indices, see
+[this documentation page](http://docs.julialang.org/en/latest/devdocs/offset-arrays.html#Arrays-with-custom-indices-1).
+
+## We know where we're going...
+
+Sometimes it's convenient to refer to a whole block of an array.  In
+Julia, if we define an array
+
+```julia
+julia> A = collect(reshape(1:30, 5, 6))
+5×6 Array{Int64,2}:
+ 1   6  11  16  21  26
+ 2   7  12  17  22  27
+ 3   8  13  18  23  28
+ 4   9  14  19  24  29
+ 5  10  15  20  25  30
+```
+
+then we can refer to a rectangular region like this:
+
+```julia
+julia> B = A[1:3, 1:4]
+3×4 Array{Int64,2}:
+ 1  6  11  16
+ 2  7  12  17
+ 3  8  13  18
+```
+
+We can represent the same concept graphically using images:
+
+```julia
+using TestImages, Images
+
+img = testimage("mandrill")
+imgeye = img[35:99,130:220]
+imgnose = img[331:445,145:350]
+```
+
+If we display these three images, we get the following:
+
+![mandrill](https://github.com/JuliaLang/julialang.github.com/blob/master/blog/_posts/offsetarrays_figures/mandrill.png?raw=true)
+![mandrill](https://github.com/JuliaLang/julialang.github.com/blob/master/blog/_posts/offsetarrays_figures/mandrill_eye.png?raw=true)
+![mandrill](https://github.com/JuliaLang/julialang.github.com/blob/master/blog/_posts/offsetarrays_figures/mandrill_nose.png?raw=true)
+
+In summary, it's very straightforward to pick a particular region of
+an array and extract it for further analysis.  Like the Talking Heads
+song "Road to Nowhere," we know where we're going...
```

Comment:
Question for you, Tim (related to what follows).  Let's say I wanted to extract a region from an image into a buffer.  I'm going to do a lot of expensive calculations on that region, so it makes sense to copy that data into a buffer instead of using a view into the original image.  Let's also say that the buffer will be a fixed size, but the target location will change once I'm done processing one object.  For example, I grab the left eye, calculate, then grab the right eye and calculate.  

Can I mutate the ranges used to define the offset array (keeping the array size fixed) so I don't have to fill up the garbage collector with a whole bunch of buffers?